File: model/Business.ts
import { Model, Relation, Query } from '@nozbe/watermelondb';
import { field, readonly, date, relation, children } from '@nozbe/watermelondb/decorators';
import type User from './User';
import type Location from './Location';
import type BusinessAssumption from './BusinessAssumption';
import type StressTest from './StressTest';
import type Pilot from './Pilot';
import type Evidence from './Evidence';
import type Decision from './Decision';
import type FinanceAssessment from './FinanceAssessment';

export default class Business extends Model {
    @field('server_revision') serverRevision!: number;
    static table = 'businesses';
    static associations = {
        users: { type: 'belongs_to', key: 'user_id' },
        locations: { type: 'belongs_to', key: 'location_id' },
        business_assumptions: { type: 'has_many', foreignKey: 'business_id' },
        stress_tests: { type: 'has_many', foreignKey: 'business_id' },
        pilots: { type: 'has_many', foreignKey: 'business_id' },
        evidence: { type: 'has_many', foreignKey: 'business_id' },
        decisions: { type: 'has_many', foreignKey: 'business_id' },
        finance_assessments: { type: 'has_many', foreignKey: 'business_id' },
    } as const;

    @field('user_id') userId!: string;
    @field('location_id') locationId!: string;
    @field('business_name') businessName!: string;
    @field('business_category') businessCategory!: string;
    @field('description') description!: string;
    @field('status') status!: string; // planning, testing, validated, financed
    @readonly @date('created_at') createdAt!: number;
    @date('updated_at') updatedAt!: number;

    @relation('users', 'user_id') user!: Relation<User>;
    @relation('locations', 'location_id') location!: Relation<Location>;
    @children('business_assumptions') assumptions!: Query<BusinessAssumption>;
    @children('stress_tests') stressTests!: Query<StressTest>;
    @children('pilots') pilots!: Query<Pilot>;
    @children('evidence') evidence!: Query<Evidence>;
    @children('decisions') decisions!: Query<Decision>;
    @children('finance_assessments') financeAssessments!: Query<FinanceAssessment>;
}

File: backend/app/models/business.py
from datetime import datetime
from typing import List, Optional
from sqlalchemy import String, Text, DateTime, ForeignKey, Index
from sqlalchemy.orm import Mapped, mapped_column, relationship
from app.db.base import Base, SyncableMixin, generate_uuid_str, utc_now


class Business(Base, SyncableMixin):
    __tablename__ = "businesses"

    id: Mapped[str] = mapped_column(
        String(64), primary_key=True, default=generate_uuid_str
    )
    user_id: Mapped[str] = mapped_column(
        String(64), ForeignKey("users.id", ondelete="CASCADE"), nullable=False, index=True
    )
    location_id: Mapped[Optional[str]] = mapped_column(
        String(64), ForeignKey("locations.id", ondelete="SET NULL"), nullable=True, index=True
    )
    business_name: Mapped[str] = mapped_column(String(200), nullable=False, index=True)
    business_category: Mapped[str] = mapped_column(String(100), nullable=False, index=True)
    description: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    status: Mapped[str] = mapped_column(
        String(50), default="DRAFT", nullable=False, index=True
    )

    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=utc_now, nullable=False
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=utc_now, onupdate=utc_now, nullable=False
    )

    # Relationships
    user: Mapped["User"] = relationship("User", back_populates="businesses")
    location: Mapped[Optional["Location"]] = relationship("Location", back_populates="businesses")
    market_data: Mapped[List["MarketData"]] = relationship(
        "MarketData", back_populates="business", cascade="all, delete-orphan"
    )
    assumptions: Mapped[List["BusinessAssumption"]] = relationship(
        "BusinessAssumption", back_populates="business", cascade="all, delete-orphan"
    )
    stress_tests: Mapped[List["StressTest"]] = relationship(
        "StressTest", back_populates="business", cascade="all, delete-orphan"
    )
    pilots: Mapped[List["Pilot"]] = relationship(
        "Pilot", back_populates="business", cascade="all, delete-orphan"
    )
    finance_assessments: Mapped[List["FinanceAssessment"]] = relationship(
        "FinanceAssessment", back_populates="business", cascade="all, delete-orphan"
    )
    evidence: Mapped[List["Evidence"]] = relationship(
        "Evidence", back_populates="business", cascade="all, delete-orphan"
    )
    decisions: Mapped[List["Decision"]] = relationship(
        "Decision", back_populates="business", cascade="all, delete-orphan"
    )

    __table_args__ = (
        Index("ix_businesses_user_category", "user_id", "business_category"),
    )
